include/alpaka/queue/cuda_hip/QueueUniformCudaHipRtBase.hpp

```diff
@@ -142,22 +165,133 @@ namespace alpaka
             }
         };
 
-        //! The CUDA/HIP RT blocking queue thread wait trait specialization.
+        //! The CUDA/HIP RT queue thread wait trait specialization.
         //!
         //! Blocks execution of the calling thread until the queue has finished processing all previously requested
         //! tasks (kernels, data copies, ...)
-        template<>
-        struct CurrentThreadWaitFor<uniform_cuda_hip::detail::QueueUniformCudaHipRtBase>
+        template<bool TBlocking>
+        struct CurrentThreadWaitFor<uniform_cuda_hip::detail::QueueUniformCudaHipRtBase<TBlocking>>
         {
             ALPAKA_FN_HOST static auto currentThreadWaitFor(
-                uniform_cuda_hip::detail::QueueUniformCudaHipRtBase const& queue) -> void
+                uniform_cuda_hip::detail::QueueUniformCudaHipRtBase<TBlocking> const& queue) -> void
             {
                 ALPAKA_DEBUG_MINIMAL_LOG_SCOPE;
 
                 // Sync is allowed even for queues on non current device.
                 ALPAKA_UNIFORM_CUDA_HIP_RT_CHECK(ALPAKA_API_PREFIX(StreamSynchronize)(queue.getNativeHandle()));
             }
         };
+
+        //! The CUDA/HIP RT queue enqueue trait specialization.
+        template<bool TBlocking, typename TTask>
+        struct Enqueue<uniform_cuda_hip::detail::QueueUniformCudaHipRtBase<TBlocking>, TTask>
+        {
+            enum class CallbackState
+            {
+                enqueued,
+                notified,
+                finished,
+            };
+
+            struct CallbackSynchronizationData : public std::enable_shared_from_this<CallbackSynchronizationData>
+            {
+                std::mutex m_mutex;
+                std::condition_variable m_event;
+                CallbackState state = CallbackState::enqueued;
```

Comment:
Please either use the `m_` prefix or don't use it. I prefer to not have it here.

Reply:
I didn't change it from what it was like before...
Anyway, what is the preferred naming scheme ?
I'd rather not have simply `std::mutex mutex` right above.

Reply:
We haven't yet defined this in our style guide, but our work-in-progress clang-tidy file mandates the `m_` prefix.

Reply:
According to our [.clang-tidy](https://github.com/ComputationalRadiationPhysics/contributing/blob/master/formatting-tools/.clang-tidy) we have a leading `m_`. But I also remember a discussion that we don't need to for aggregates. @j-stephan?

Reply:
`m_state` it is, then.